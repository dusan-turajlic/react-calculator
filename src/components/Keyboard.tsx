import React from "react";
import Key from './Key';

interface KeyboardProps {
    onAnyKeyPress?( event: React.MouseEvent, value: string ): void
}

class Keyboard extends React.Component<KeyboardProps> {
    render() {
        return (
            <div className="container">
                <div className="columns">
                    <Key onPress={ this.props.onAnyKeyPress } value="1"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="2"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="3"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="/" classNames={ [ 'bg-gray' ] }/>
                </div>
                <div className="columns">
                    <Key onPress={ this.props.onAnyKeyPress } value="4"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="5"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="6"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="*" classNames={ [ 'bg-gray' ] }/>
                </div>
                <div className="columns">
                    <Key onPress={ this.props.onAnyKeyPress } value="7"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="8"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="9"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="-" classNames={ [ 'bg-gray' ] }/>
                </div>
                <div className="columns">
                    <Key onPress={ this.props.onAnyKeyPress } value="."/>
                    <Key onPress={ this.props.onAnyKeyPress } value="0"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="ce"/>
                    <Key onPress={ this.props.onAnyKeyPress } value="+" classNames={ [ 'bg-gray' ] }/>
                </div>
                <div className="columns">
                    <Key onPress={ this.props.onAnyKeyPress } value="=" colSize="" classNames={ [ 'bg-warning' ] }/>
                </div>
            </div>
        );
    }
}

export default Keyboard;
