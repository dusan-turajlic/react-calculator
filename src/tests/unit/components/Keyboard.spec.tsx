import React, { DOMElement } from "react";
import { render, fireEvent } from '@testing-library/react';
import Keyboard from "../../../components/Keyboard";

describe( 'Keyboard.spec', () => {
    it( 'renders correctly', function () {
        const { container } = render( <Keyboard /> );
        expect( container ).toMatchSnapshot();
    } );

    it( 'passing callback will get triggered on any key button press', function () {
        const onAnyKeyPressCallback = jest.fn();
        const keyboard = render( <Keyboard onAnyKeyPress={onAnyKeyPressCallback} /> );
        const buttons = [
            keyboard.getByText( '1' ),
            keyboard.getByText( '2' ),
            keyboard.getByText( '3' ),
            keyboard.getByText( '4' ),
            keyboard.getByText( '5' ),
            keyboard.getByText( '6' ),
            keyboard.getByText( '7' ),
            keyboard.getByText( '8' ),
            keyboard.getByText( '9' ),
            keyboard.getByText( '0' ),
            keyboard.getByText( '.' ),
            keyboard.getByText( 'ce' ),
            keyboard.getByText( '/' ),
            keyboard.getByText( '*' ),
            keyboard.getByText( '-' ),
            keyboard.getByText( '+' ),
            keyboard.getByText( '=' ),
        ];

        buttons.forEach( ( button ) => {
            fireEvent.click( button );
        } );

        expect( onAnyKeyPressCallback ).toHaveBeenCalledTimes( 17 );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '1' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '2' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '3' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '/' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '4' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '5' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '6' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '*' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '7' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '8' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '9' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '-' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '.' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '0' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), 'ce' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '+' );
        expect( onAnyKeyPressCallback ).toHaveBeenCalledWith( expect.any( Object ), '=' );
    } );
} );
