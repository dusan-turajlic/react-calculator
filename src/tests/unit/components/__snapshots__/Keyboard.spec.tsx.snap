// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard.spec renders correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="columns"
    >
      <button
        class="column col-3 "
      >
        1
      </button>
      <button
        class="column col-3 "
      >
        2
      </button>
      <button
        class="column col-3 "
      >
        3
      </button>
      <button
        class="column col-3 bg-gray"
      >
        /
      </button>
    </div>
    <div
      class="columns"
    >
      <button
        class="column col-3 "
      >
        4
      </button>
      <button
        class="column col-3 "
      >
        5
      </button>
      <button
        class="column col-3 "
      >
        6
      </button>
      <button
        class="column col-3 bg-gray"
      >
        *
      </button>
    </div>
    <div
      class="columns"
    >
      <button
        class="column col-3 "
      >
        7
      </button>
      <button
        class="column col-3 "
      >
        8
      </button>
      <button
        class="column col-3 "
      >
        9
      </button>
      <button
        class="column col-3 bg-gray"
      >
        -
      </button>
    </div>
    <div
      class="columns"
    >
      <button
        class="column col-3 "
      >
        .
      </button>
      <button
        class="column col-3 "
      >
        0
      </button>
      <button
        class="column col-3 "
      >
        ce
      </button>
      <button
        class="column col-3 bg-gray"
      >
        +
      </button>
    </div>
    <div
      class="columns"
    >
      <button
        class="column  bg-warning"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
